naga/tests/validation.rs

```diff
@@ -311,3 +311,116 @@ fn main() {{
         assert_eq!(err.emit_to_string(&source), expected_err);
     }
 }
+
+#[cfg(feature = "wgsl-in")]
+#[test]
+fn incompatible_interpolation_and_sampling_types() {
+    // NOTE: Things we expect to actually compile are in the `interpolate` snapshot test.
+    use itertools::Itertools;
+
+    let invalid_shader_module = |interpolation_and_sampling| {
+        let (interpolation, sampling) = interpolation_and_sampling;
+
+        let valid = matches!(
+            (interpolation, sampling),
+            (_, None)
+                | (
+                    naga::Interpolation::Perspective | naga::Interpolation::Linear,
+                    Some(
+                        naga::Sampling::Center | naga::Sampling::Centroid | naga::Sampling::Sample
+                    ),
+                )
+                | (
+                    naga::Interpolation::Flat,
+                    Some(naga::Sampling::First | naga::Sampling::Either)
+                )
+        );
+
+        if valid {
+            None
+        } else {
+            // NOTE: If you have to add variants below, make sure to add them to the
+            // `cartesian_product` call below!
+            let interpolation_str = match interpolation {
+                naga::Interpolation::Flat => "flat",
+                naga::Interpolation::Linear => "linear",
+                naga::Interpolation::Perspective => "perspective",
+            };
+            let sampling_str = match sampling {
+                None => String::new(),
+                Some(sampling) => format!(
+                    ", {}",
+                    match sampling {
+                        naga::Sampling::First => "first",
+                        naga::Sampling::Either => "either",
+                        naga::Sampling::Center => "center",
+                        naga::Sampling::Centroid => "centroid",
+                        naga::Sampling::Sample => "sample",
+                    }
+                ),
+            };
+            let member_type = match interpolation {
+                naga::Interpolation::Perspective | naga::Interpolation::Linear => "f32",
+                naga::Interpolation::Flat => "f32",
```

Comment:
I originally intended to use `u32` for some variety here, but this is `f32` like the other branch. Oops!

Reply:
Fixed with latest force-push to 0591c9a5.